File: backend/app/schemas/goal.py
from datetime import datetime

from pydantic import BaseModel, Field


class GoalBase(BaseModel):
    title: str = Field(..., min_length=1, max_length=200)
    description: str | None = Field(default=None, max_length=2000)
    target_date: datetime
    status: str = Field(..., min_length=1, max_length=50)
    progress: int = Field(..., ge=0, le=100)


class GoalCreate(GoalBase):
    pass


class GoalUpdate(GoalBase):
    pass


class GoalResponse(GoalBase):
    id: int
    user_id: int
    created_at: datetime
    updated_at: datetime

    class Config:
        from_attributes = True

File: frontend/src/types/goal.ts
export type GoalStatus = 'not_started' | 'in_progress' | 'completed'

export type Goal = {
  id: number
  user_id: number
  title: string
  description: string | null
  target_date: string
  status: GoalStatus
  progress: number
  created_at: string
  updated_at: string
}

export type GoalPayload = {
  title: string
  description: string | null
  target_date: string
  status: GoalStatus
  progress: number
}

export type GoalQueryParams = {
  skip?: number
  limit?: number
  search?: string
  status?: GoalStatus | ''
  sort_by?: 'created_at' | 'updated_at' | 'target_date' | 'title' | 'status' | 'progress'
  sort_dir?: 'asc' | 'desc'
}
